src/base/static/components/form-field/index.js

```diff
@@ -1,354 +1,114 @@
-  onChange(fieldName, fieldValue) {
-    this.props.onChange(
-      fieldName,
-      fieldValue,
-      this.validator.validate(
-        fieldValue,
-        this.props.places,
-        this.props.landmarks
-      ),
-      this.validator.message
-    );
-  }
-
-  getValidator(fieldConfig) {
-    if (fieldConfig.type === constants.SUBMIT_FIELD_TYPENAME) {
-      return {
-        validate: mayHaveAnyValue,
-        message: "",
-      };
-    } else if (fieldConfig.type === constants.MAP_DRAWING_TOOLBAR_TYPENAME) {
-      return {
-        validate: mustHaveSomeValue,
-        message: messages.missingGeometry,
-      };
-    } else if (fieldConfig.type === constants.CUSTOM_URL_TOOLBAR_TYPENAME) {
-      return {
-        validate: mustHaveUniqueUrl,
-        message: messages.duplicateUrl,
-      };
-    } else {
-      return {
-        validate: fieldConfig.optional ? mayHaveAnyValue : mustHaveSomeValue,
-        message: messages.missingRequired,
-      };
-    }
-  }
+    this.fieldDefinition = fieldDefinitions[this.props.fieldConfig.type];
 
-  getInitialFieldState(fieldConfig) {
     // "autofill" is a better term than "autocomplete" for this feature.
     // TODO: Update this throughout the codebase.
-    let autofillValue = fieldConfig.autocomplete
-      ? Util.getAutocompleteValue(fieldConfig.name)
+    const autofillValue = this.props.fieldConfig.autocomplete
+      ? Util.getAutocompleteValue(this.props.fieldConfig.name)
       : null;
-
-    fieldConfig.hasAutofill = !!autofillValue;
-
-    let fieldValue;
-    switch (fieldConfig.type) {
-      case constants.BIG_TOGGLE_FIELD_TYPENAME:
-        fieldValue =
-          autofillValue ||
-          fieldConfig.default_value ||
-          fieldConfig.content[1].value; // "off" position of the toggle
-        break;
-      case constants.BIG_CHECKBOX_FIELD_TYPENAME:
-        fieldValue =
-          fromJS(autofillValue) ||
-          fromJS(fieldConfig.default_value) ||
-          ImmutableList();
-        break;
-      case constants.PUBLISH_CONTROL_TOOLBAR_TYPENAME:
-        fieldValue = autofillValue || fieldConfig.default_value;
-        break;
-      case constants.MAP_DRAWING_TOOLBAR_TYPENAME:
-        fieldValue = null;
-        break;
-      default:
-        fieldValue = autofillValue || fieldConfig.default_value || "";
-        break;
-    }
-
-    const fieldObj = ImmutableMap()
+    this.props.fieldConfig.hasAutofill = !!autofillValue;
+    const initialFieldValue = this.fieldDefinition.getInitialValue(
+      autofillValue,
+      this.props.fieldConfig.default_value,
+      this.props.fieldConfig
+    );
+    const initialFieldState = ImmutableMap()
       .set(constants.FIELD_STATE_INITIALIZED_KEY, true)
-      .set(constants.FIELD_STATE_VALUE_KEY, fieldValue)
-      .set(constants.FIELD_STATE_FIELD_TYPE_KEY, fieldConfig.type)
+      .set(constants.FIELD_STATE_VALUE_KEY, initialFieldValue)
+      .set(constants.FIELD_STATE_FIELD_TYPE_KEY, this.props.fieldConfig.type)
```

Comment:
What's the purpose of this `FIELD_STATE_INITIALIZED_KEY`? It doesn't seem to be getting used.

Same goes for `FIELD_STATE_FIELD_TYPE_KEY`.

Reply:
Ah good catch-- those are left over from a previous implementation.